server/server.js

```diff
@@ -1,15 +1,157 @@
 
 // Store and retrieve your videos from here
 // If you want, you can copy "exampleresponse.json" into here to have some data to work with
-let videos = [];
+let videos = [
+  {
+    id: 523523,
+    title: "Never Gonna Give You Up",
+    url: "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
+    rating: 23,
+  },
+  {
+    id: 523427,
+    title: "The Coding Train",
+    url: "https://www.youtube.com/watch?v=HerCR8bw_GE",
+    rating: 230,
+  },
+  {
+    id: 82653,
+    title: "Mac & Cheese | Basics with Babish",
+    url: "https://www.youtube.com/watch?v=FUeyrEN14Rk",
+    rating: 2111,
+  },
+  {
+    id: 858566,
+    title: "Videos for Cats to Watch - 8 Hour Bird Bonanza",
+    url: "https://www.youtube.com/watch?v=xbs7FT7dXYc",
+    rating: 11,
+  },
+  {
+    id: 453538,
+    title:
+      "The Complete London 2012 Opening Ceremony | London 2012 Olympic Games",
+    url: "https://www.youtube.com/watch?v=4As0e4de-rI",
+    rating: 3211,
+  },
+  {
+    id: 283634,
+    title: "Learn Unity - Beginner's Game Development Course",
+    url: "https://www.youtube.com/watch?v=gB1F9G0JXOo",
+    rating: 211,
+  },
+  {
+    id: 562824,
+    title: "Cracking Enigma in 2021 - Computerphile",
+    url: "https://www.youtube.com/watch?v=RzWB5jL5RX0",
+    rating: 111,
+  },
+  {
+    id: 442452,
+    title: "Coding Adventure: Chess AI",
+    url: "https://www.youtube.com/watch?v=U4ogK0MIzqk",
+    rating: 671,
+  },
+  {
+    id: 536363,
+    title: "Coding Adventure: Ant and Slime Simulations",
+    url: "https://www.youtube.com/watch?v=X-iSQQgOd1A",
+    rating: 76,
+  },
+  {
+    id: 323445,
+    title: "Why the Tour de France is so brutal",
+    url: "https://www.youtube.com/watch?v=ZacOS8NBK6U",
+    rating: 73,
+  },
+];
 
 // GET "/"
 app.get("/", (req, res) => {
-  // Delete this line after you've confirmed your server is running
-  res.send({ express: "Your Backend Service is Running" });
+  const { order } = req.query;
+  let orderedVideos = [...videos];
+  if (order === "asc") {
+    orderedVideos.sort((a, b) => a.rating - b.rating);
+  } else {
+    orderedVideos.sort((a, b) => b.rating - a.rating);
+  }
+
+  res.json(orderedVideos);
+});
+
+// GET "/{id}"
+app.get("/:id", (req, res) => {
+  const id = parseInt(req.params.id);
+
+  const video = videos.find((v) => v.id === id);
+
+  if (!video) {
+    res.status(404).json({
+      result: "failure",
+      message: "Video not found",
+    });
+    return;
+  }
+
+  res.json(video);
+});
+
+app.post("/", (req, res) => {
+  // adding the checks here after suggested by Jonathan so front-end can be cleaner
+  const { title, url } = req.body;
+
+  if (
+    !title ||
+    !url ||
+    !url.match(/^(https?:\/\/)?(www\.)?youtube\.com\/watch\?v=[\w-]+(&\S+)?$/)
+  ) {
+    res.send({
+      result: "failure",
+      message: "Video could not be saved",
+    });
+  }
+
+  let newId = 0;
+  while (videos.some((video) => video.id === newId)) {
```

Comment:
There are couple of really interesting things about this way of choosing an ID:
1. You may re-use them - if you add a video, then delete it, a new video may get its ID - can you think of any places this may be problematic?
2. (Advanced topic, and I'm very happy to chat about this if you want) It involves lots of searching through the videos list, and it will slow down a lot as you get more videos. Let's imagine you had 1000 videos with IDs 0..999 - you're going to look through all 1000 videos, 1000 times, so you're going to do 1 million "does this video already have this ID" checks. Can you think of a way that we could do this faster?